module/zfs/dsl_bookmark.c

```diff
@@ -116,57 +120,177 @@ dsl_bookmark_lookup(dsl_pool_t *dp, const char *fullname,
+ *   ({bookmark,snapshot}_namecheck)
+ * - all newbookmarks are in the same pool
+ * - all newbookmarks have unique names
+ * - FIXME: need to check that newbookmark is never used as source in any other?
+ *
+ * Note that this does not check for existence of the pool or source bookmarks.
+ *
+ * Returns 0 if the nvlist adheres to above schema.
+ * Returns -1 if it doesn't.
+ */
+int
+dsl_bookmark_create_nvl_validate(nvlist_t *bmarks)
+{
+	char *first, *source;
+	size_t first_len;
+
+	first = NULL;
+	for (nvpair_t *pair = nvlist_next_nvpair(bmarks, NULL);
+	    pair != NULL; pair = nvlist_next_nvpair(bmarks, pair)) {
+		int is_bmark, is_snap;
+
+		char *bmark = nvpair_name(pair);
+
+		if (bookmark_namecheck(bmark, NULL, NULL) != 0)
+			return (-1);
+
+		/* structure check: values must be snapshots XOR bookmarks */
+		if (nvpair_value_string(pair, &source) != 0)
+			return (-1);
+		is_bmark = bookmark_namecheck(source, NULL, NULL) == 0;
+		is_snap = snapshot_namecheck(source, NULL, NULL) == 0;
+		if (!is_bmark && !is_snap) {
+			return (-1);
+		}
+
+		/* same pool check */
+		if (first == NULL) {
+			char *cp = strpbrk(bmark, "/#");
+			if (cp == NULL)
+				return (-1);
+			first = bmark;
+			first_len = cp - bmark;
+		}
+
+		if (strncmp(first, bmark, first_len) != 0)
+			return (-1);
+
+		if (*(bmark + first_len) != '/')
+			return (-1);
 
-typedef struct dsl_bookmark_create_arg {
-	nvlist_t *dbca_bmarks;
-	nvlist_t *dbca_errors;
-} dsl_bookmark_create_arg_t;
+		/* unique newbookmark names; todo: O(n^2) */
+		for (nvpair_t *pair2 = nvlist_next_nvpair(bmarks, pair);
+		    pair2 != NULL; pair2 = nvlist_next_nvpair(bmarks, pair2)) {
+			if (strcmp(nvpair_name(pair), nvpair_name(pair2)) == 0)
+				return (-1);
+		}
+
+	}
+	return (0);
+}
 
 static int
-dsl_bookmark_create_check_impl(dsl_dataset_t *snapds, const char *bookmark_name,
-    dmu_tx_t *tx)
+dsl_bookmark_create_check_impl_snap(dsl_pool_t *dp,
+    dsl_dataset_t *newbm_ds, const char *newbm_shortname,
+    dsl_dataset_t *source_snap_ds)
 {
-	dsl_pool_t *dp = dmu_tx_pool(tx);
-	dsl_dataset_t *bmark_fs;
-	char *shortname;
+	/*
+	 * Verify that the source snapshot is an earlier point in
+	 * newbm_ds's timeline (source may be origin of newbm_ds)
+	 */
+	if (!dsl_dataset_is_before(newbm_ds, source_snap_ds, 0)) {
+		return (EINVAL);
+	}
+	return (0);
+}
+
+static int
+dsl_bookmark_create_check_impl_book(dsl_pool_t *dp,
+    dsl_dataset_t *newbm_ds, const char *source)
+{
+	ASSERT3B(bookmark_namecheck(source, NULL, NULL), ==, 0);
+
+	zfs_bookmark_phys_t source_bm_phys;
 	int error;
-	zfs_bookmark_phys_t bmark_phys = { 0 };
 
-	if (!snapds->ds_is_snapshot)
-		return (SET_ERROR(EINVAL));
+	/*
+	 * Verify that the source bookmark exists and is an earlier point in
+	 * newbm_ds's timeline (newbm_ds's origin may be a snap of source's ds)
+	 */
+	error = dsl_bookmark_lookup(dp, source, newbm_ds, &source_bm_phys);
+	if (error != 0)
+		return (error);
+
+	return (0);
+}
```

Comment:
I think we could get rid of these 2 functions and put their code in-line.  They're only a couple lines of code.